diff --git a/config.def.h b/config.def.h
index 08ff707..3800f63 100644
--- a/config.def.h
+++ b/config.def.h
@@ -25,7 +25,8 @@ static Bool allowgeolocation = TRUE;
 
 #define SETPROP(p, q) { \
 	.v = (char *[]){ "/bin/sh", "-c", \
-		"prop=\"`xprop -id $2 $0 | cut -d '\"' -f 2 | xargs -0 printf %b | dmenu`\" &&" \
+		"prop=\"`(xprop -id $2 $0 | cut -d '\"' -f 2 | xargs -0 printf %b && "\
+                "cat ~/.surf/bookmarks) | dmenu`\" &&" \
 		"xprop -id $2 -f $1 8s -set $1 \"$prop\"", \
 		p, q, winid, NULL \
 	} \
@@ -41,6 +42,12 @@ static Bool allowgeolocation = TRUE;
 	} \
 }
 
+#define BM_ADD { .v = (char *[]){ "/bin/sh", "-c", \
+  "(echo `xprop -id $0 _SURF_URI | cut -d '\"' -f 2` && "\
+  "cat ~/.surf/bookmarks) | awk '!seen[$0]++' > ~/.surf/bookmarks_new && "\
+  "mv ~/.surf/bookmarks_new ~/.surf/bookmarks", \
+  winid, NULL } }
+
 #define MODKEY GDK_CONTROL_MASK
 
 /* hotkeys */
@@ -72,6 +79,7 @@ static Key keys[] = {
     { MODKEY,               GDK_space,       scroll_v,   { .i = +10000 } },
     { MODKEY,               GDK_i,           scroll_h,   { .i = +1 } },
     { MODKEY,               GDK_u,           scroll_h,   { .i = -1 } },
+    { MODKEY,               GDK_b,      spawn,      BM_ADD },
 
     { 0,                    GDK_F11,    fullscreen, { 0 } },
     { 0,                    GDK_Escape, stop,       { 0 } },
@@ -93,4 +101,3 @@ static Key keys[] = {
     { MODKEY|GDK_SHIFT_MASK,GDK_b,      togglescrollbars,{ 0 } },
     { MODKEY|GDK_SHIFT_MASK,GDK_g,      togglegeolocation, { 0 } },
 };
-
